### ISSUE TITLE
AOPC JDBC Driver Bug

### ISSUE THREAD

**Issue by renedlog** [neo4j-contrib/neo4j-apoc-procedures#1421]
The Simba BigQuery JDBC driver does not work with the current APOC 4.0.0.4 in Neo4J 4.0.0. Although it worked before and is still flawlessly working with plain java.
A simple example that should work:
```
CALL apoc.load.jdbc('jdbc:bigquery://https://www.googleapis.com/bigquery/v2:443;ProjectId=project-id;OAuthType=0;OAuthServiceAcctEmail=user@project-id.iam.gserviceaccount.com;OAuthPvtKeyPath=C:/Users/user/google.json;', 'select 1 as a')
YIELD row
RETURN row
``` 

The query above returns the following stack trace:

```
2020-02-24 14:25:24.544+0000 ERROR Cannot execute SQL statement `'SELECT 1'`.
Error:
[Simba][JDBC](10220) Driver does not support this optional feature. [Simba][JDBC](10220) Driver does not support this optional feature.
java.sql.SQLFeatureNotSupportedException: [Simba][JDBC](10220) Driver does not support this optional feature.
	at com.simba.googlebigquery.exceptions.ExceptionConverter.toSQLException(Unknown Source)
	at com.simba.googlebigquery.jdbc.common.SConnection.setAutoCommit(Unknown Source)
	at apoc.load.Jdbc.executeQuery(Jdbc.java:79)
	at apoc.load.Jdbc.jdbc(Jdbc.java:62)
	at org.neo4j.kernel.impl.proc.GeneratedProcedure_jdbc9432664354800.apply(Unknown Source)
	at org.neo4j.procedure.impl.ProcedureRegistry.callProcedure(ProcedureRegistry.java:208)
	at org.neo4j.procedure.impl.GlobalProceduresRegistry.callProcedure(GlobalProceduresRegistry.java:323)
	at org.neo4j.kernel.impl.newapi.AllStoreHolder.callProcedure(AllStoreHolder.java:975)
	at org.neo4j.kernel.impl.newapi.AllStoreHolder.procedureCallRead(AllStoreHolder.java:878)
	at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$.$anonfun$callReadOnlyProcedure$2(CallSupport.scala:50)
	at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$.callProcedure(CallSupport.scala:97)
	at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$.callReadOnlyProcedure(CallSupport.scala:52)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext.callReadOnlyProcedure(TransactionBoundQueryContext.scala:822)
	at org.neo4j.cypher.internal.planning.ExceptionTranslatingQueryContext.$anonfun$callReadOnlyProcedure$1(ExceptionTranslatingQueryContext.scala:180)
	at org.neo4j.cypher.internal.planning.ExceptionTranslationSupport.translateException(ExceptionTranslationSupport.scala:33)
	at org.neo4j.cypher.internal.planning.ExceptionTranslationSupport.translateException$(ExceptionTranslationSupport.scala:32)
	at org.neo4j.cypher.internal.planning.ExceptionTranslatingQueryContext.translateException(ExceptionTranslatingQueryContext.scala:40)
	at org.neo4j.cypher.internal.planning.ExceptionTranslationSupport.translateIterator(ExceptionTranslationSupport.scala:48)
	at org.neo4j.cypher.internal.planning.ExceptionTranslationSupport.translateIterator$(ExceptionTranslationSupport.scala:47)
	at org.neo4j.cypher.internal.planning.ExceptionTranslatingQueryContext.translateIterator(ExceptionTranslatingQueryContext.scala:40)
	at org.neo4j.cypher.internal.planning.ExceptionTranslatingQueryContext.callReadOnlyProcedure(ExceptionTranslatingQueryContext.scala:180)
	at org.neo4j.cypher.internal.runtime.LazyReadOnlyCallMode.callProcedure(ProcedureCallMode.scala:48)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.ProcedureCallPipe.call(ProcedureCallPipe.scala:87)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.ProcedureCallPipe.$anonfun$internalCreateResultsByAppending$1(ProcedureCallPipe.scala:73)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:480)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:486)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.SlottedPipeHeadOperator$OTask.operate(SlottedPipeOperator.scala:111)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.OperatorWithInterpretedDBHitsProfiling.operateWithProfile(SlottedPipeOperator.scala:51)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.OperatorWithInterpretedDBHitsProfiling.operateWithProfile$(SlottedPipeOperator.scala:43)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.SlottedPipeHeadOperator$OTask.operateWithProfile(SlottedPipeOperator.scala:93)
	at org.neo4j.cypher.internal.runtime.pipelined.PipelineTask.executeOperators(PipelineTask.scala:58)
	at org.neo4j.cypher.internal.runtime.pipelined.PipelineTask.executeWorkUnit(PipelineTask.scala:48)
	at org.neo4j.cypher.internal.runtime.pipelined.Worker.executeTask(Worker.scala:128)
	at org.neo4j.cypher.internal.runtime.pipelined.Worker.workOnQuery(Worker.scala:84)
	at org.neo4j.cypher.internal.runtime.pipelined.execution.CallingThreadExecutingQuery.request(CallingThreadExecutingQuery.scala:30)
	at org.neo4j.cypher.internal.PipelinedRuntime$PipelinedRuntimeResult.request(PipelinedRuntime.scala:326)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:88)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.request(ClosingExecutionResult.scala:135)
	at org.neo4j.bolt.runtime.AbstractCypherAdapterStream.handleRecords(AbstractCypherAdapterStream.java:105)
	at org.neo4j.bolt.v3.messaging.ResultHandler.onPullRecords(ResultHandler.java:41)
	at org.neo4j.bolt.v4.messaging.PullResultConsumer.consume(PullResultConsumer.java:42)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State.consumeResult(TransactionStateMachine.java:511)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State$1.streamResult(TransactionStateMachine.java:264)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine.streamResult(TransactionStateMachine.java:92)
	at org.neo4j.bolt.v4.runtime.AutoCommitState.processStreamResultMessage(AutoCommitState.java:41)
	at org.neo4j.bolt.v4.runtime.AbstractStreamingState.processUnsafe(AbstractStreamingState.java:49)
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48)
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:143)
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:91)
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:90)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:151)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:240)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:175)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:165)
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:212)
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$2(ExecutorBoltScheduler.java:195)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

As said above there is no problem using other tools. So it seems there is a bug in APOC 4.0.0.4

**Comment by sarmbruster** [comment]
The referred line in `apoc.load.Jdbc#79` is 
```
connection.setAutoCommit(false);
```
That line has been adding back in Oct '19 to fix a postgres related performance issue.

@renedlog do you know if the simba jdbc driver allows disabling autoCommmit ? The error message indicates it doesn't. 

We could easily wrap that line into a try-catch block and ignore that exception (of course dumping it to log file before).

**Comment by renedlog** [comment]
@sarmbruster right. Seems the driver does not support disabled autoCommit. How exactly I am supposed to enable auto-commit if using Cypher? Seems there is no documentation about it.

Anyway weird that APOC is going this way just for Postgres... there are more drivers not supporting disabled auto-commit than the other way around. That's the main reason auto-commit in almost all tools and programming languages I used so far.

**Comment by conker84** [comment]
> We could easily wrap that line into a try-catch block and ignore that exception (of course dumping it to log file before).

@sarmbruster we can add an option `autoCommit: <default, false>` and the leave to the user
the power to choose the best option for him?

**Comment by moxious** [comment]
@renedlog at Neo4j we happen to have some contacts with the underlying vendor of the JDBC driver you're using.  A response I received from them:

It seems like there was a change in APOC that started setting auto commit to false, starting transaction, on the BigQuery driver but the BigQuery driver does not support transactions. Google BigQuery only supports auto commit. In any case, I have given a heads up to our Google BigQuery team. However, the best course of action would be for the user who reported this issue to file a support ticket for this issue in our support portal.

**Comment by jexp** [comment]
@conker84 please do implement this.

**Comment by conker84** [comment]
closed via 8e9cf0f
